### PR TITLE
Feature compute

### PR DESCRIPTION
**Pull request by 161563**
created compute functions for command buffer
Render tasks can be also be created as a compute or copy task
Changed deferred composition render task to now use the compute pipeline
Added a copy to back buffer render task, for displaying the results of compute calls

### CODE REVIEW

File: src/d3d12/d3d12_command_list.cpp [hunk 1]
```diff
@@ -196,11 +205,22 @@ namespace wr::d3d12
 		cmd_list->m_native->SetGraphicsRootConstantBufferView(root_parameter_idx, buffer->m_gpu_addresses[frame_idx]);
 	}
 
+	void BindCompute(CommandList * cmd_list, HeapResource* buffer, unsigned int root_parameter_idx, unsigned int frame_idx)
```
Comment (VZout) [inline src/d3d12/d3d12_command_list.cpp]: Rename BindCompute etc to BindComputeXXX please so we know what it is binding.

File: src/d3d12/d3d12_device.cpp [hunk 2]
```diff
@@ -194,7 +194,8 @@ namespace wr::d3d12
 			//internal::EnableDXRFallback();
 		}
 
-		TRY_M(D3D12CreateDevice(device->m_adapter, device->m_feature_level, IID_PPV_ARGS(&device->m_native)),
+		HRESULT hr;
+		TRY_M(hr = D3D12CreateDevice(device->m_adapter, device->m_feature_level, IID_PPV_ARGS(&device->m_native)),
```
Comment (VZout) [inline src/d3d12/d3d12_device.cpp]: Revert this please. This is for debugging purposes.

File: src/d3d12/d3d12_enums.hpp [hunk 3]
```diff
@@ -138,6 +138,7 @@ namespace wr
 		DEPTH_WRITE = (int)D3D12_RESOURCE_STATE_DEPTH_WRITE,
 		DEPTH_READ = (int)D3D12_RESOURCE_STATE_DEPTH_READ,
 		INDIRECT_ARGUMENT = (int)D3D12_RESOURCE_STATE_INDIRECT_ARGUMENT,
+		NON_PIXEL_SHADER_RESOURCE = (int)D3D12_RESOURCE_STATE_NON_PIXEL_SHADER_RESOURCE,
```
Comment (VZout) [inline src/d3d12/d3d12_enums.hpp]: Duplicate now I believe.

File: src/render_tasks/d3d12_deferred_render_target_copy.hpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	struct DeferredRenderTargetCopyTaskData
+	{
+		d3d12::RenderTarget* out_deferred_composition_rt;
+	};
+	using DeferredRenderTargetCopyRenderTask_t = RenderTask<DeferredRenderTargetCopyTaskData>;
+
+	namespace internal
+	{
+
+		inline void RecordDrawCommands(D3D12RenderSystem& render_system, d3d12::CommandList* cmd_list, d3d12::HeapResource* camera_cb, DeferredRenderTargetCopyTaskData const & data, unsigned int frame_idx)
```
Comment (VZout) [inline src/render_tasks/d3d12_deferred_render_target_copy.hpp]: remove if you don't use bundles.

File: src/render_tasks/d3d12_deferred_render_target_copy.hpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+		{
+			
+		}
+
+		inline void SetupDeferredTask(RenderSystem & render_system, DeferredRenderTargetCopyRenderTask_t & task, DeferredRenderTargetCopyTaskData & data)
```
Comment (VZout) [inline src/render_tasks/d3d12_deferred_render_target_copy.hpp]: Rename functions to match the actual render task please.

File: src/render_tasks/d3d12_deferred_render_target_copy.hpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+				src.pResource = data.out_deferred_composition_rt->m_render_targets[0];
+				src.Type = D3D12_TEXTURE_COPY_TYPE::D3D12_TEXTURE_COPY_TYPE_SUBRESOURCE_INDEX;
+				src.SubresourceIndex = 0;
+
+				cmd_list->m_native->CopyTextureRegion(&dst, 0, 0, 0, &src, nullptr);
+			}
+		}
+
+
+		inline void ResizeDeferredTask(DeferredRenderTargetCopyRenderTask_t & task, DeferredRenderTargetCopyTaskData& data, std::uint32_t width, std::uint32_t height)
```
Comment (VZout) [inline src/render_tasks/d3d12_deferred_render_target_copy.hpp]: Remove unused functions

File: CMakeLists.txt [hunk 7]
```diff
@@ -127,7 +127,7 @@ set_target_properties(IrrXML PROPERTIES FOLDER ThirdParty)
 set_target_properties(UpdateAssimpLibsDebugSymbolsAndDLLs PROPERTIES FOLDER ThirdParty)
 set_target_properties(zlib PROPERTIES FOLDER ThirdParty)
 set_target_properties(zlibstatic PROPERTIES FOLDER ThirdParty)
-set_target_properties(unit PROPERTIES FOLDER ThirdParty)
```
Comment (VZout) [inline CMakeLists.txt]: Revert this please.

File: resources/shaders/deferred_composition.hlsl [hunk 8]
```diff
@@ -91,19 +92,24 @@ float3 shade_pixel(float3 vpos, float3 V, float3 albedo, float3 normal)
...
+	output.GetDimensions(screen_size.x, screen_size.y);
+	float2 uv = float2(dispatch_thread_id.x / screen_size.x, 1.f - (dispatch_thread_id.y / screen_size.y));
+
+	float2 screen_coord = int2(dispatch_thread_id.x, screen_size.y - dispatch_thread_id.y);
 
 	// GBuffer contents
-	const float3 albedo = gbuffer_albedo.Sample(s0, uv).xyz;
-	const float3 normal = gbuffer_normal.Sample(s0, uv).xyz;
-	const float depth_f = gbuffer_depth.Sample(s0, uv).r;
+	const float3 albedo = gbuffer_albedo[screen_coord].xyz;
```
Comment (VZout) [inline resources/shaders/deferred_composition.hlsl]: Prefer sample over [] operator.

File: src/d3d12/d3d12_command_list.cpp [hunk 9]
```diff
@@ -160,6 +160,15 @@ namespace wr::d3d12
 		cmd_list->m_native->SetDescriptorHeaps(num, n_heaps.data());
 	}
 
+	void BindComputePipeline(CommandList* cmd_list, PipelineState * pipeline_state)
+	{
+		if (pipeline_state->m_type == PipelineType::COMPUTE_PIPELINE)
```
Comment (VZout) [inline src/d3d12/d3d12_command_list.cpp]: Atleast warn if this happens (`LOGW`). Or just let it crash. The debug layer will warn what went wrong.

File: resources/shaders/deferred_composition.hlsl (hunk 8 above)
Comment (161563) [reply resources/shaders/deferred_composition.hlsl]: Sample causes ugly aliasing issues.

File: CMakeLists.txt (hunk 7 above)
Comment (161563) [reply CMakeLists.txt]: Uncommented it.

File: src/d3d12/d3d12_command_list.cpp (hunk 9 above)
Comment (161563) [reply src/d3d12/d3d12_command_list.cpp]: Now logging a warning instead of doing nothing

**Pull request merged**